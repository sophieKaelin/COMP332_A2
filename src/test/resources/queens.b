//
// This file is part of COMP332 Assignment 2/3 2019.
//
// weBCPL, a retro BCPL to WebAssembly compiler.
//
// © 2019, Dominic Verity and Anthony Sloane, Macquarie University.
//         All rights reserved.
// © 2005, Martin Richards
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Solution to the n-Queens problem from Martin Richards BCPL distribution.
//

GLOBAL 
$(
	count: 200
	all: 201
$)

LET try(ld, row, rd) BE
	TEST row = all THEN
		count := count + 1
	ELSE
    $(
		LET poss = all & ~(ld | row | rd)
		UNTIL poss = 0 DO
        $(
			LET p = poss & -poss
			poss := poss - p
			try(ld + p << 1, row + p, rd + p >> 1)
		$)
	$)

LET start() = VALOF
$(
	all := 1
	FOR i = 1 TO 12 DO
    $(
		count := 0
		try(0, 0, 0)
		writef("%i2-Queens problem has %i5 solutions*n", i, count)
		all := 2 * all + 1
	$)
	RESULTIS 0
$)
