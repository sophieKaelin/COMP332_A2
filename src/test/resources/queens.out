Program(
    Vector(
        GlobalDecl(
            Vector(
                GlobalEntry(IdnDef("count"), Some(IntExp(200))),
                GlobalEntry(IdnDef("all"), Some(IntExp(201))))),
        LetDecl(
            Vector(
                LetProcClause(
                    IdnDef("try"),
                    Vector(IdnDef("ld"), IdnDef("row"), IdnDef("rd")),
                    TestThenElseStmt(
                        EqualExp(IdnExp(IdnUse("row")), IdnExp(IdnUse("all"))),
                        AssignStmt(
                            Vector(IdnExp(IdnUse("count"))),
                            Vector(PlusExp(IdnExp(IdnUse("count")), IntExp(1)))),
                        Block(
                            Vector(
                                LetDecl(
                                    Vector(
                                        LetVarClause(
                                            Vector(IdnDef("poss")),
                                            Vector(
                                                AndExp(
                                                    IdnExp(IdnUse("all")),
                                                    NotExp(
                                                        OrExp(
                                                            OrExp(
                                                                IdnExp(IdnUse("ld")),
                                                                IdnExp(IdnUse("row"))),
                                                            IdnExp(IdnUse("rd")))))))))),
                            Vector(
                                UntilDoStmt(
                                    EqualExp(IdnExp(IdnUse("poss")), IntExp(0)),
                                    Block(
                                        Vector(
                                            LetDecl(
                                                Vector(
                                                    LetVarClause(
                                                        Vector(IdnDef("p")),
                                                        Vector(
                                                            AndExp(
                                                                IdnExp(IdnUse("poss")),
                                                                NegExp(IdnExp(IdnUse("poss"))))))))),
                                        Vector(
                                            AssignStmt(
                                                Vector(IdnExp(IdnUse("poss"))),
                                                Vector(
                                                    MinusExp(
                                                        IdnExp(IdnUse("poss")),
                                                        IdnExp(IdnUse("p"))))),
                                            CallStmt(
                                                CallExp(
                                                    IdnExp(IdnUse("try")),
                                                    Vector(
                                                        ShiftLeftExp(
                                                            PlusExp(
                                                                IdnExp(IdnUse("ld")),
                                                                IdnExp(IdnUse("p"))),
                                                            IntExp(1)),
                                                        PlusExp(
                                                            IdnExp(IdnUse("row")),
                                                            IdnExp(IdnUse("p"))),
                                                        ShiftRightExp(
                                                            PlusExp(
                                                                IdnExp(IdnUse("rd")),
                                                                IdnExp(IdnUse("p"))),
                                                            IntExp(1)))))))))))))),
        LetDecl(
            Vector(
                LetFnClause(
                    IdnDef("start"),
                    Vector(),
                    ValofExp(
                        Block(
                            Vector(),
                            Vector(
                                AssignStmt(Vector(IdnExp(IdnUse("all"))), Vector(IntExp(1))),
                                ForStmt(
                                    IdnDef("i"),
                                    IntExp(1),
                                    IntExp(12),
                                    None,
                                    Block(
                                        Vector(),
                                        Vector(
                                            AssignStmt(
                                                Vector(IdnExp(IdnUse("count"))),
                                                Vector(IntExp(0))),
                                            CallStmt(
                                                CallExp(
                                                    IdnExp(IdnUse("try")),
                                                    Vector(IntExp(0), IntExp(0), IntExp(0)))),
                                            CallStmt(
                                                CallExp(
                                                    IdnExp(IdnUse("writef")),
                                                    Vector(
                                                        StringExp(
                                                            Vector(
                                                                37,
                                                                105,
                                                                50,
                                                                45,
                                                                81,
                                                                117,
                                                                101,
                                                                101,
                                                                110,
                                                                115,
                                                                32,
                                                                112,
                                                                114,
                                                                111,
                                                                98,
                                                                108,
                                                                101,
                                                                109,
                                                                32,
                                                                104,
                                                                97,
                                                                115,
                                                                32,
                                                                37,
                                                                105,
                                                                53,
                                                                32,
                                                                115,
                                                                111,
                                                                108,
                                                                117,
                                                                116,
                                                                105,
                                                                111,
                                                                110,
                                                                115,
                                                                10)),
                                                        IdnExp(IdnUse("i")),
                                                        IdnExp(IdnUse("count"))))),
                                            AssignStmt(
                                                Vector(IdnExp(IdnUse("all"))),
                                                Vector(
                                                    PlusExp(
                                                        StarExp(IntExp(2), IdnExp(IdnUse("all"))),
                                                        IntExp(1))))))),
                                ResultIsStmt(IntExp(0))))))))))
